src/loan/SimplePSM.sol

```diff
@@ -0,0 +1,102 @@
+contract SimplePSM is CoreRef {
+    using SafeERC20 for ERC20;
+
+    /// @notice reference to the ProfitManager contract
+    address public immutable profitManager;
+
+    /// @notice reference to the RateLimitedMinter contract to mint CREDIT
+    address public immutable rlcm;
+
+    /// @notice reference to the CreditToken contract
+    address public immutable credit;
+
+    /// @notice reference to the peg token contract
+    address public immutable token;
+
+    /// @notice multiplier for decimals correction, e.g. 1e12 for a token
+    /// with 6 decimals (because CREDIT has 18 decimals)
+    uint256 public immutable decimalCorrection;
+
+    /// @notice event emitted upon a redemption
+    event Redeem(address indexed to, uint256 amountIn, uint256 amountOut);
+    /// @notice event emitted when credit gets minted
+    event Mint(address indexed to, uint256 amountIn, uint256 amountOut);
+
+    constructor(
+        address _core,
+        address _profitManager,
+        address _rlcm,
+        address _credit,
+        address _token
+    ) CoreRef(_core) {
+        profitManager = _profitManager;
+        rlcm = _rlcm;
+        credit = _credit;
+        token = _token;
+
+        uint256 decimals = uint256(ERC20(_token).decimals());
+        decimalCorrection = 10**(18 - decimals);
+    }
+
+    /// @notice calculate the amount of CREDIT out for a given `amountIn` of underlying
+    function getMintAmountOut(uint256 amountIn) public view returns (uint256) {
+        uint256 creditMultiplier = ProfitManager(profitManager).creditMultiplier();
+        return amountIn * decimalCorrection * 1e18 / creditMultiplier;
+    }
+
+    /// @notice calculate the amount of underlying out for a given `amountIn` of CREDIT
+    function getRedeemAmountOut(uint256 amountIn) public view returns (uint256) {
+        uint256 creditMultiplier = ProfitManager(profitManager).creditMultiplier();
+        return amountIn * creditMultiplier / 1e18 / decimalCorrection;
+    }
+
+    /// @notice mint `amountOut` CREDIT to address `to` for `amountIn` underlying tokens
+    /// @dev see getMintAmountOut() to pre-calculate amount out
+    function mint(
+        address to,
+        uint256 amountIn
+    ) external whenNotPaused returns (uint256 amountOut) {
+        amountOut = getMintAmountOut(amountIn);
+        ERC20(token).safeTransferFrom(msg.sender, address(this), amountIn);
+        RateLimitedMinter(rlcm).mint(to, amountOut);
+        emit Mint(to, amountIn, amountOut);
+    }
+
+    /// @notice redeem `amountIn` CREDIT for `amountOut` underlying tokens and send to address `to`
+    /// @dev see getRedeemAmountOut() to pre-calculate amount out
+    function redeem(
+        address to,
+        uint256 amountIn
+    ) external returns (uint256 amountOut) {
+        amountOut = getRedeemAmountOut(amountIn);
+        CreditToken(credit).transferFrom(msg.sender, address(this), amountIn);
+        CreditToken(credit).burn(amountIn);
```

Comment:
`burnFrom` would do the same thing in less calls and save a warm SSTORE + SLOAD + call.